src/state/searchState.ts

```diff
@@ -116,74 +141,81 @@ export class SearchState {
+   *
    * NOTE: This method does not take the search offset into account
    */
   public getNextSearchMatchRange(
     vimState: VimState,
     fromPosition: Position,
-    direction = SearchDirection.Forward
-  ): { range: Range; index: number } | undefined {
+    direction = SearchDirection.Forward,
+    relativeIndex = 0
+  ): IndexedRange | undefined {
     const matchRanges = this.recalculateSearchRanges(vimState);
 
     if (matchRanges.length === 0) {
       return undefined;
     }
 
     const effectiveDirection = (direction * this.direction) as SearchDirection;
+    let index: number | undefined;
 
     if (effectiveDirection === SearchDirection.Forward) {
-      for (const [index, range] of matchRanges.entries()) {
-        if (range.start.isAfter(fromPosition)) {
-          return {
-            range,
-            index,
-          };
+      for (let i = 0; i < matchRanges.length; i++) {
+        if (
+          (this.offset?.apply(matchRanges[i]) ?? matchRanges[i].start).compareTo(fromPosition) > 0
+        ) {
+          index = i;
+          break;
         }
       }
-      // We've hit the bottom of the file. Wrap around if configured to do so, or return undefined.
-      if (configuration.wrapscan) {
-        const range = matchRanges[0];
-        return {
-          range,
-          index: 0,
-        };
-      } else {
-        return undefined;
-      }
     } else {
-      for (const [index, range] of matchRanges.slice(0).reverse().entries()) {
-        if (range.end.isBeforeOrEqual(fromPosition)) {
-          return {
-            range,
-            index: matchRanges.length - index - 1,
-          };
+      for (let i = matchRanges.length - 1; i >= 0; i--) {
+        if (
+          (this.offset?.apply(matchRanges[i]) ?? matchRanges[i].start).compareTo(fromPosition) < 0
+        ) {
+          index = i;
+          break;
         }
       }
+    }
 
-      // We've hit the top of the file. Wrap around if configured to do so, or return undefined.
+    if (index === undefined) {
+      // We've hit the top/bottom of the file. Wrap around if configured to do so, or return undefined.
       if (configuration.wrapscan) {
-        const range = matchRanges[matchRanges.length - 1];
-        return {
-          range,
-          index: matchRanges.length - 1,
-        };
+        index = effectiveDirection === SearchDirection.Forward ? 0 : matchRanges.length - 1;
       } else {
         return undefined;
       }
     }
+
+    // index of the first match now stored in variable index.
+    // Offsetting it by relativeIndex in the appropriate direction gets the index of the desired match.
+    index += effectiveDirection * relativeIndex;
+
+    if (0 <= index && index < matchRanges.length) {
+      return { index, range: matchRanges[index] };
+    }
+
+    // We've hit the top/bottom of the file. Wrap around (possibly many times) if configured to do so, or return undefined
```

Comment:
By `(possibly many times)`, you mean via `mod`, right? We won't loop through the matches multiple times

Reply:
That's right, I was just talking about the apparent behavior